File: dev/python/automated_imgur_upload/imgur.py
import json
import os
import time
import webbrowser
from datetime import datetime, timedelta

import requests


def load_config():
    with open("automated_imgur_upload/config/config.json", "r") as file:
        return json.load(file)


def load_tokens():
    token_file = "automated_imgur_upload/config/tokens.json"
    if os.path.exists(token_file):
        with open(token_file, "r") as file:
            return json.load(file)
    return None


def save_tokens(tokens):
    token_file = "automated_imgur_upload/config/tokens.json"
    os.makedirs(os.path.dirname(token_file), exist_ok=True)

    tokens["expires_at"] = (
        datetime.now() + timedelta(seconds=tokens["expires_in"])
    ).isoformat()

    with open(token_file, "w") as file:
        json.dump(tokens, file, indent=2)


def is_token_valid(tokens):
    if not tokens or "expires_at" not in tokens:
        return False

    expires_at = datetime.fromisoformat(tokens["expires_at"])
    # Consider token expired 5 minutes before actual expiration
    return datetime.now() < expires_at - timedelta(minutes=5)


def refresh_access_token(client_id, client_secret, refresh_token):
    token_url = "https://api.imgur.com/oauth2/token"
    data = {
        "refresh_token": refresh_token,
        "client_id": client_id,
        "client_secret": client_secret,
        "grant_type": "refresh_token",
    }

    response = requests.post(token_url, data=data)
    if response.status_code == 200:
        return response.json()
    return None


def get_authorization_url(client_id):
    return f"https://api.imgur.com/oauth2/authorize?client_id={client_id}&response_type=pin&state=APPLICATION_STATE"


def get_tokens(client_id, client_secret, pin):
    token_url = "https://api.imgur.com/oauth2/token"
    data = {
        "client_id": client_id,
        "client_secret": client_secret,
        "grant_type": "pin",
        "pin": pin,
    }
    response = requests.post(token_url, data=data)
    return response.json()


def get_valid_access_token(client_id, client_secret):
    stored_tokens = load_tokens()

    if stored_tokens and is_token_valid(stored_tokens):
        print("✅ Using existing valid access token")
        return stored_tokens["access_token"]

    if stored_tokens and "refresh_token" in stored_tokens:
        print("🔄 Refreshing access token...")
        new_tokens = refresh_access_token(
            client_id, client_secret, stored_tokens["refresh_token"]
        )

        if new_tokens and new_tokens.get("success", True):
            print("✅ Successfully refreshed access token")
            save_tokens(new_tokens)
            return new_tokens["access_token"]
        else:
            print("❌ Failed to refresh token, need new authorization")

    print("🔐 Need fresh authorization...")
    authorization_url = get_authorization_url(client_id)
    print(f"Please go to this URL and authorize access: {authorization_url}")
    webbrowser.open(authorization_url)

    pin = input("Enter the pin obtained from the authorization URL: ")
    tokens = get_tokens(client_id, client_secret, pin)

    if tokens.get("success", True):
        print("✅ Successfully obtained new tokens")
        save_tokens(tokens)
        return tokens["access_token"]
    else:
        raise Exception(f"Failed to get tokens: {tokens}")


def upload_image(image_key, image_repository, access_token, album_id):
    image_info = image_repository[image_key]
    relative_subpath = image_info["path"]
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.abspath(os.path.join(script_dir, "..", "..", ".."))
    image_path = os.path.join(project_root, relative_subpath)

    if not os.path.isfile(image_path):
        print(f"❌ File not found at {image_path}")
        return {"success": False, "error": f"File not found: {image_path}"}

    title = image_info["title"]
    description = image_info["description"] or title
    url = "https://api.imgur.com/3/upload"
    headers = {"Authorization": f"Bearer {access_token}"}

    with open(image_path, "rb") as image_file:
        files = {"image": image_file}
        data = {"album": album_id, "title": title, "description": description}
        response = requests.post(url, headers=headers, files=files, data=data)

    try:
        return response.json()
    except ValueError:
        return {
            "success": False,
            "status_code": response.status_code,
            "raw_text": response.text,
        }


def upload_images_in_order(image_order, image_repository, access_token, album_id):
    for image_key in image_order:
        upload_response = upload_image(
            image_key, image_repository, access_token, album_id
        )

        if upload_response.get("success", False):
            print(f"✅ Uploaded {image_key}")
        else:
            print(f"❌ Failed {image_key}: {upload_response}")

        time.sleep(0.15)


def delete_image(deletehash, access_token):
    url = f"https://api.imgur.com/3/image/{deletehash}"
    headers = {"Authorization": f"Bearer {access_token}"}
    response = requests.delete(url, headers=headers)
    return response.json()


def delete_all_images_from_album(album_id, access_token):
    url = f"https://api.imgur.com/3/album/{album_id}/images"
    headers = {"Authorization": f"Bearer {access_token}"}
    response = requests.get(url, headers=headers)

    if response.status_code != 200:
        print(f"❌ Failed to get album images: {response.status_code}")
        return

    images = response.json()["data"]

    if images:
        print(f"🗑️  Deleting {len(images)} images from album...")
        for image in images:
            deletehash = image["deletehash"]
            delete_image(deletehash, access_token)
    else:
        print("ℹ️ No images to delete from album")


def print_credit_limit(access_token):
    url = "https://api.imgur.com/3/credits"
    headers = {"Authorization": f"Bearer {access_token}"}
    response = requests.get(url, headers=headers)

    if response.status_code != 200:
        print(f"❌ Failed to get credits: {response.status_code}")
        return

    credits = response.json()["data"]
    print(f"User Limit: {credits['UserLimit']}")
    print(f"User Remaining: {credits['UserRemaining']}")
    print(f"User Reset: {credits['UserReset']}")
    print(f"Client Limit: {credits['ClientLimit']}")
    print(f"Client Remaining: {credits['ClientRemaining']}")


def imgur():
    config = load_config()
    client_id = config["client_id"]
    client_secret = config["client_secret"]
    album_id = "YoxqS7A"

    access_token = get_valid_access_token(client_id, client_secret)

    image_order = config["image_order"]

    with open("automated_imgur_upload/config/image_info.json", "r") as file:
        image_info = json.load(file)

    delete_all_images_from_album(album_id, access_token)
    print("✅ Finished Deletion")

    upload_images_in_order(image_order, image_info, access_token, album_id)
    print("✅ Finished Uploading")

    print()
    print("🎉 Success!!")
    print()
    print_credit_limit(access_token)
    print()


if __name__ == "__main__":
    imgur()

File: packages/gsheets2img/src/imgur.ts
import fs from "fs"
import path from "path"
import { createInterface } from "readline"
import open from "open"

import config from "@/config/config.json"

interface Config {
  client_id: string
  client_secret: string
  image_order: string[]
}

interface Tokens {
  access_token: string
  refresh_token: string
  expires_in: number
  expires_at?: string
}

interface ImageInfo {
  path: string
  title: string
  description?: string
}

const loadConfig = async (): Promise<Config> => {
  return config
}

const loadTokens = async (): Promise<Tokens | null> => {
  const tokenFile = "../config/tokens.json"
  const file = Bun.file(tokenFile)
  if (!(await file.exists())) {
    return null
  }
  const content = await file.text()
  return JSON.parse(content)
}

const saveTokens = async (tokens: Tokens): Promise<void> => {
  const tokenFile = "../config/tokens.json"
  fs.mkdirSync(path.dirname(tokenFile), { recursive: true })
  tokens.expires_at = new Date(
    Date.now() + tokens.expires_in * 1000,
  ).toISOString()
  await Bun.write(tokenFile, JSON.stringify(tokens, null, 2))
}

const loadImageInfo = async (): Promise<Record<string, ImageInfo>> => {
  const content = await Bun.file("../config/image_info.json").text()
  return JSON.parse(content)
}

const isTokenValid = (tokens: Tokens | null): boolean => {
  if (!tokens || !tokens.expires_at) {
    return false
  }
  const expiresAt = new Date(tokens.expires_at)
  return new Date() < new Date(expiresAt.getTime() - 5 * 60 * 1000)
}

const refreshAccessToken = async (
  clientId: string,
  clientSecret: string,
  refreshToken: string,
): Promise<any> => {
  const tokenUrl = "https://api.imgur.com/oauth2/token"
  const data = {
    refresh_token: refreshToken,
    client_id: clientId,
    client_secret: clientSecret,
    grant_type: "refresh_token",
  }
  const response = await fetch(tokenUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(data).toString(),
  })
  return response.json()
}

const getAuthorizationUrl = (clientId: string): string => {
  return `https://api.imgur.com/oauth2/authorize?client_id=${clientId}&response_type=pin&state=APPLICATION_STATE`
}

const getTokens = async (
  clientId: string,
  clientSecret: string,
  pin: string,
): Promise<any> => {
  const tokenUrl = "https://api.imgur.com/oauth2/token"
  const data = {
    client_id: clientId,
    client_secret: clientSecret,
    grant_type: "pin",
    pin: pin,
  }
  const response = await fetch(tokenUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/x-www-form-urlencoded",
    },
    body: new URLSearchParams(data).toString(),
  })
  return response.json()
}

const getValidAccessToken = async (
  clientId: string,
  clientSecret: string,
): Promise<string> => {
  const storedTokens = await loadTokens()

  if (storedTokens && isTokenValid(storedTokens)) {
    console.log("✅ Using existing valid access token")
    return storedTokens.access_token
  }

  if (storedTokens && storedTokens.refresh_token) {
    console.log("🔄 Refreshing access token...")
    const newTokens = await refreshAccessToken(
      clientId,
      clientSecret,
      storedTokens.refresh_token,
    )

    if (newTokens && newTokens.success !== false) {
      console.log("✅ Successfully refreshed access token")
      await saveTokens(newTokens)
      return newTokens.access_token
    } else {
      console.log("❌ Failed to refresh token, need new authorization")
    }
  }

  console.log("🔐 Need fresh authorization...")
  const authorizationUrl = getAuthorizationUrl(clientId)
  console.log(`Please go to this URL and authorize access: ${authorizationUrl}`)
  await open(authorizationUrl)

  const rl = createInterface({
    input: process.stdin,
    output: process.stdout,
  })
  const pin = await new Promise<string>((resolve) => {
    rl.question(
      "Enter the pin obtained from the authorization URL: ",
      (answer) => {
        rl.close()
        resolve(answer)
      },
    )
  })

  const tokens = await getTokens(clientId, clientSecret, pin)

  if (tokens.success !== false) {
    console.log("✅ Successfully obtained new tokens")
    await saveTokens(tokens)
    return tokens.access_token
  } else {
    throw new Error(`Failed to get tokens: ${JSON.stringify(tokens)}`)
  }
}

const uploadImage = async (
  imageKey: string,
  imageRepository: Record<string, ImageInfo>,
  accessToken: string,
  albumId: string,
): Promise<any> => {
  const imageInfo = imageRepository[imageKey]
  const relativeSubpath = imageInfo.path
  const scriptDir = path.dirname(new URL(import.meta.url).pathname)
  const projectRoot = path.resolve(scriptDir, "..", "..", "frontend")
  const imagePath = path.join(projectRoot, relativeSubpath)

  if (!(await Bun.file(imagePath).exists())) {
    console.log(`❌ File not found at ${imagePath}`)
    return { success: false, error: `File not found: ${imagePath}` }
  }

  const title = imageInfo.title
  const description = imageInfo.description || title
  const url = "https://api.imgur.com/3/upload"
  const headers = { Authorization: `Bearer ${accessToken}` }

  const imageFile = new Blob([await Bun.file(imagePath).bytes()])
  const formData = new FormData()
  formData.append("image", imageFile)
  formData.append("album", albumId)
  formData.append("title", title)
  formData.append("description", description)

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: headers,
      body: formData,
    })
    return response.json()
  } catch (error: any) {
    return {
      success: false,
      status_code: error.response?.status,
      raw_text: error.response?.data,
    }
  }
}

const uploadImagesInOrder = async (
  imageOrder: string[],
  imageRepository: Record<string, ImageInfo>,
  accessToken: string,
  albumId: string,
): Promise<void> => {
  for (const imageKey of imageOrder) {
    const uploadResponse = await uploadImage(
      imageKey,
      imageRepository,
      accessToken,
      albumId,
    )

    if (uploadResponse.success !== false) {
      console.log(`✅ Uploaded ${imageKey}`)
    } else {
      console.log(`❌ Failed ${imageKey}: ${JSON.stringify(uploadResponse)}`)
    }

    await new Promise((resolve) => setTimeout(resolve, 150))
  }
}

const deleteImage = async (
  deletehash: string,
  accessToken: string,
): Promise<any> => {
  const url = `https://api.imgur.com/3/image/${deletehash}`
  const headers = { Authorization: `Bearer ${accessToken}` }
  const response = await fetch(url, { method: "DELETE", headers })
  return response.json()
}

const deleteAllImagesFromAlbum = async (
  albumId: string,
  accessToken: string,
): Promise<void> => {
  const url = `https://api.imgur.com/3/album/${albumId}/images`
  const headers = { Authorization: `Bearer ${accessToken}` }
  const response = await fetch(url, { headers })

  if (!response.ok) {
    console.log(`❌ Failed to get album images: ${response.status}`)
    return
  }

  const data = await response.json()
  const images = data.data

  if (images.length > 0) {
    console.log(`🗑️  Deleting ${images.length} images from album...`)
    for (const image of images) {
      await deleteImage(image.deletehash, accessToken)
    }
  } else {
    console.log("ℹ️ No images to delete from album")
  }
}

const printCreditLimit = async (accessToken: string): Promise<void> => {
  const url = "https://api.imgur.com/3/credits"
  const headers = { Authorization: `Bearer ${accessToken}` }
  const response = await fetch(url, { headers })

  if (!response.ok) {
    console.log(`❌ Failed to get credits: ${response.status}`)
    return
  }

  const data = await response.json()
  const credits = data.data
  console.log(`User Limit: ${credits.UserLimit}`)
  console.log(`User Remaining: ${credits.UserRemaining}`)
  console.log(`User Reset: ${credits.UserReset}`)
  console.log(`Client Limit: ${credits.ClientLimit}`)
  console.log(`Client Remaining: ${credits.ClientRemaining}`)
}

export const imgur = async (): Promise<void> => {
  const config = await loadConfig()
  const clientId = config.client_id
  const clientSecret = config.client_secret
  const albumId = "YoxqS7A"

  const accessToken = await getValidAccessToken(clientId, clientSecret)

  const imageOrder = config.image_order
  const imageInfo = await loadImageInfo()

  await deleteAllImagesFromAlbum(albumId, accessToken)
  console.log("✅ Finished Deletion")

  await uploadImagesInOrder(imageOrder, imageInfo, accessToken, albumId)
  console.log("✅ Finished Uploading")

  console.log()
  console.log("🎉 Success!!")
  console.log()
  await printCreditLimit(accessToken)
  console.log()
}

if (import.meta.url === `file://${process.argv[1]}`) {
  imgur().catch(console.error)
}
